File: project1/src/task3.py
import json
import os
import sys
import time

from pyspark import SparkConf, SparkContext


def isValidFiles(input_file):
    if not os.path.isfile(input_file):
        return False
    return True


def default(sc, input_file, n_partitions, n_arg):
    start = time.time()

    lines = sc.textFile(input_file).coalesce(n_partitions).cache()

    rdd = lines.map(json.loads).map(lambda x: (x["business_id"], 1))

    num_partitions = rdd.getNumPartitions()

    n_items = rdd.glom().map(len).collect()

    business_sum = rdd.reduceByKey(lambda x, y: x + y)

    popular_businesses = business_sum.filter(lambda x: x[1] > n_arg)

    results = popular_businesses.collect()

    # Convert list of tuples to list of lists
    results = list(map(list, results))

    end = time.time()

    print("Default Time ----> ", end - start)

    return (num_partitions, n_items, results)

def customized(sc, input_file, n_partitions, n_arg):
    start = time.time()

    lines = sc.textFile(input_file)

    rdd = lines.map(json.loads).map(lambda x: (x["business_id"], 1)).partitionBy(n_partitions, lambda x: hash(x)).cache()

    num_partitions = rdd.getNumPartitions()

    n_items = rdd.glom().map(len).collect()

    business_sum = rdd.reduceByKey(lambda x, y: x + y, 8)

    popular_businesses = business_sum.filter(lambda x: x[1] > n_arg)

    results = popular_businesses.collect()
    results = list(map(list, results))

    end = time.time()

    print('Cutomized Time ----> ', end - start)

    return (num_partitions, n_items, results)


def main(argv):
    assert len(argv) == 5, "Script takes 5 arguments"

    # Unpack arguments
    input_file, output_file, partition_type, n_partitions, n_arg = argv

    # Check for valid file paths
    if not isValidFiles(input_file):
        print("Invalid path arguments")
        return

    n_arg = int(n_arg)
    n_partitions = int(n_partitions)
    out_dict = {}

    # Create Spark configuration and context
    conf = SparkConf().setMaster("local[*]").setAppName("Task3")
    sc = SparkContext(conf=conf)

    if partition_type == "default":
        (num_partitions, n_items, results) = default(sc, input_file, n_partitions, n_arg)
        out_dict["n_partitions"] = num_partitions
        out_dict["n_items"] = n_items
        out_dict["result"] = results
    else:
        (num_partitions, n_items, results) = customized(sc, input_file, n_partitions, n_arg)
        out_dict["n_partitions"] = num_partitions
        out_dict["n_items"] = n_items
        out_dict["result"] = results

    with open(output_file, "w") as file:
        json.dump(out_dict, file)

if __name__ == "__main__":
    main(sys.argv[1:])

File: project1_scala/hw1/src/main/scala/task3.scala
import java.io.FileWriter

import org.apache.log4j.{Level, Logger}
import org.apache.spark.{HashPartitioner, SparkContext}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.JsonDSL._

import scala.collection.mutable.ListBuffer

//noinspection DuplicatedCode
object task3 extends Serializable {
  def time[R](block: => R): R = {
    val t0 = System.nanoTime()
    val result = block
    val t1 = System.nanoTime()
    println("Elapsed time: " + (t1 - t0)/1e9d + "sec")
    result
  }

  def mapBusinessID(jsonObj: JObject): (String, Int) = {
    implicit val formats = DefaultFormats
    ((jsonObj \ "business_id").extract[String], 1)
  }

  def default(inputFile: String, numPartitions: Int, n: Int): String = {

    Logger.getLogger("org").setLevel(Level.ERROR)

    val sc = new SparkContext("local[*]", "Task3")

    val lines = sc.textFile(inputFile).coalesce(numPartitions).cache()

    val rdd = lines.map(x => parse(x).asInstanceOf[JObject]).map(mapBusinessID)

    val totalPartitions = rdd.partitions.length

    // Fetch partitions array and convert it to a list
    val nItems = rdd.glom().map(x => x.length).collect()
    val nItemsList: ListBuffer[Int] = ListBuffer()

    for (item <- nItems) {
      nItemsList += item
    }

    val businessSum = rdd.reduceByKey((x, y) => x + y)

    val popularBusinesses = businessSum.filter(x => x._2 > n)

    val results = popularBusinesses.collect()

    // Write results as [[String1, Int1], [String2, Int2]]

    var strResult = "["
    for (value<- results) {
      strResult += "[\"" + value._1 + "\", " + value._2 + "], "
    }
    strResult = strResult.substring(0, strResult.length - 2)
    strResult += "]"

    val outputDict = ("n_partitions" -> totalPartitions) ~
      ("n_items" -> nItemsList.toList) ~ ("result" -> parse(strResult))
    compact(render(outputDict))
  }

  def custom(inputFile: String, numPartitions: Int, n: Int): String = {
    Logger.getLogger("org").setLevel(Level.ERROR)

    val sc = new SparkContext("local[*]", "Task3")

    val lines = sc.textFile(inputFile)

    val rdd = lines.map(x => parse(x).asInstanceOf[JObject])
      .map(mapBusinessID)
      .partitionBy(new HashPartitioner(numPartitions)).cache()

    val totalPartitions = rdd.partitions.length

    val nItems = rdd.glom().map(x => x.length).collect()
    val nItemsList: ListBuffer[Int] = ListBuffer()

    for (item <- nItems) {
      nItemsList += item
    }

    val businessSum = rdd.reduceByKey((x, y) => x + y, 8).cache()

    val popularBusinesses = businessSum.filter(x => x._2 > n)

    val results = popularBusinesses.collect()

    // Write results as [[String1, Int1], [String2, Int2]]

    var strResult = "["
    for (value<- results) {
      strResult += "[\"" + value._1 + "\", " + value._2 + "], "
    }
    strResult = strResult.substring(0, strResult.length - 2)
    strResult += "]"

    val outputDict = ("n_partitions" -> totalPartitions) ~
      ("n_items" -> nItemsList.toList) ~ ("result" -> parse(strResult))
    compact(render(outputDict))
  }


  def main(args: Array[String]): Unit = {

    //Unpack arguments
    val Array(inputFile, outputFile, partitionType, numPartitions, n) = args
    var result = ""

    if (partitionType.contentEquals("default")) {
      result = time(default(inputFile, numPartitions.toInt, n.toInt))
    } else {
      result = time(custom(inputFile, numPartitions.toInt, n.toInt))
    }

    val writer = new FileWriter(outputFile)
    writer.write(result)
    writer.close()
  }
}
